xtra-libp2p-offer/src/current/maker.rs

```diff
@@ -102,9 +117,8 @@ impl Actor {
             .await;
     }
 
-    async fn handle_connection_dropped(&mut self, msg: endpoint::ConnectionDropped) {
-        tracing::trace!("Remove dropped connection: {:?}", msg.peer_id);
-        self.connected_peers.remove(&msg.peer_id);
+    async fn handle_no_connection(&mut self, msg: NoConnection) {
```

Comment:
🤔 I am not sure I would replace one solution with the other. 
I think if the endpoint already knows that the connection was dropped we should also react to it in the protocols that want to know. 

I would add what you proposed as a resilience fallback on top of the previous solution. Otherwise it is expected to see the warning above, and I don't think that is expected behavior.

Note: Same for all other protocols where this was changed.

Reply:
> I think if the endpoint already knows that the connection was dropped we should also react to it in the protocols that want to know.

I couldn't justify to myself the need for the connection-dropped notification in any of the places where I removed the handlers. It's not like anything dangerous can happen with this stale information. As soon as the actor tries to use the information they will get a `NoConnection` error which is functionally the same as receiving an eager notification.

For the places where I didn't remove the handling of connection-dropped notifications, the reasons are pretty specific:

- I wanted to remove it for the `identify::dialer::Actor`, but we currently have that questionable need to report the number of live connections from there. If we didn't have that I think that actor wouldn't need to know about dropped connections at all.
- The `online_status::Actor` is the only place where I think the subscription is a good idea. Although to be fair it is only used by the taker to keep track of one specific connection (the maker's), even though we will hear about all dropped connections (probably only the maker's, but still). One could argue that we might just want to add a simple query mechanism to the `Endpoint` to actively check if the connection is still alive and remove the dropped-connection notifications altogether!

> Otherwise it is expected to see the warning above, and I don't think that is expected behavior.

Nope, the warning can be (and is) skipped if we don't have a connection to the peer.

Reply:
I am not sure I like this solution, because we now can have different state concerning a peer being connected in different actors. 

This might be resilient for these specific actors, but I think the original idea was that the endpoint is in charge of managing connected peers. So far we always got notified about connected peers from the endpoint, i.e. we passed the state of the endpoint on to subscribers. Now these actors change this state independent of the state in the endpoint. I am not saying this is wrong, but I am not sure this is what we want.

Reply:
> So far we always got notified about connected peers from the endpoint, i.e. we passed the state of the endpoint on to subscribers. Now these actors change this state independent of the state in the endpoint.

But it is not independent of the state in the endpoint! It is the endpoint that tells them about an actor no longer being connected, but only when they ask.

If we want to eagerly synchronise these states then we can reintroduce the notifications, but I don't see the need. In fact, I would like to eventually remove the feature altogether. But that might be too much.

Reply:
> It is the endpoint that tells them about an actor no longer being connected, but only when they ask.

I am not sure I can follow. Until this PR we subscribed to state changes of the endpoint, so the endpoint did notify subscribed protocol actors about peer disconnection. 

> If we want to eagerly synchronise these states then we can reintroduce the notifications, but I don't see the need. 

Fair enough, I think the solution you came up with works, it might just not be as efficient as it could be, because we might still try to open a protocol substream with a peer even if the endpoint already knows that the peer idisconneced. Acceptable I suppose.

> In fact, I would like to eventually remove the feature altogether. But that might be too much.

You mean the feature of protocol actors subscribing to state changes of connect/disconnect in the endpoint? Initially we did not have this feature, but that lead to having to query the endpoint about peer information for all protocols where we need all peers. In the long run it might just be more efficient (and resilient) to change these protocols to be polled by the `taker` instead of the `maker` pushing the information. Then, we also don't need `ping` from `maker` to `taker` anymore.